rfaas/include/rfaas/connection.hpp

```diff
@@ -13,24 +13,104 @@
 
 namespace rfaas {
 
+  template <typename Library>
   struct manager_connection {
+
+    using RDMAActive_t = typename rdmalib::rdmalib_traits<Library>::RDMAActive;
+    using RecvBuffer_t = typename rdmalib::rdmalib_traits<Library>::RecvBuffer;
+    using Connection_t = typename rdmalib::rdmalib_traits<Library>::Connection;
+    using ScatterGatherElement_t = typename rdmalib::rdmalib_traits<Library>::ScatterGatherElement;
+
     std::string _address;
     int _port;
-    rdmalib::Buffer<char> _submit_buffer;
-    rdmalib::RDMAActive _active;
-    rdmalib::RecvBuffer _rcv_buffer;
-    rdmalib::Buffer<rdmalib::AllocationRequest> _allocation_buffer;
+    rdmalib::Buffer<char, Library> _submit_buffer;
+    RDMAActive_t _active;
+    RecvBuffer_t _rcv_buffer;
+    rdmalib::Buffer<rdmalib::AllocationRequest, Library> _allocation_buffer;
     int _max_inline_data;
 
     manager_connection(std::string address, int port, int rcv_buf, int max_inline_data);
 
-    rdmalib::Connection & connection();
+    Connection_t & connection();
     rdmalib::AllocationRequest & request();
     bool connect();
     void disconnect();
     bool submit();
   };
 
+  template <typename Library>
+  manager_connection<Library>::manager_connection(std::string address, int port,
+      int rcv_buf, int max_inline_data):
+    _address(address),
+    _port(port),
+    _active(_address, _port, rcv_buf),
+    _rcv_buffer(rcv_buf),
+    _allocation_buffer(rcv_buf + 1),
+    _max_inline_data(max_inline_data)
+  {
+    _active.allocate();
+  }
+
+  template <typename Library>
+  bool manager_connection<Library>::connect()
+  {
+    SPDLOG_DEBUG("Connecting to manager at {}:{}", _address, _port);
+    bool ret = _active.connect();
+    if(!ret) {
+      spdlog::error("Couldn't connect to manager at {}:{}", _address, _port);
+      return false;
+    }
+    #ifdef USE_LIBFABRIC
+    _allocation_buffer.register_memory(_active.pd(), FI_WRITE | FI_REMOTE_WRITE);
+    #else
+    _allocation_buffer.register_memory(_active.pd(), IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE);
```

Comment:
Shouldn't this just use the new refactored traits?

Reply:
Yes, I must have missed this one.